From: OpenWrt EHR330 support <openwrt-devel@lists.openwrt.org>
Subject: [PATCH] mediatek: add BL31-backed FIT decryption

Add the MediaTek AES-256-CBC backend used by encrypted FIT images. Payloads
are decrypted in place through the BL31 firmware-decryption SMC service and
key material is erased before Linux starts.

---
 arch/arm/lib/bootm.c           |  7 +++++++
 arch/arm/mach-mediatek/Kconfig | 15 +++++++++++++++
 boot/Makefile                  |  2 ++
 boot/image-cipher.c            | 15 ++++++++++++++-
 include/u-boot/aes.h           | 13 +++++++++++++
 5 files changed, 51 insertions(+), 1 deletion(-)

--- a/arch/arm/mach-mediatek/Kconfig
+++ b/arch/arm/mach-mediatek/Kconfig
@@ -172,4 +172,17 @@ config RESET_BUTTON_SETTLE_DELAY
 config RESET_BUTTON_SETTLE_DELAY
 	int "Delay to wait for button to settle"
 	default 0
+
+config MTK_FW_ENCRYPT
+	bool "MediaTek encrypted FIT firmware support"
+	select FIT_CIPHER
+	help
+	  Enable decryption of encrypted FIT payloads using a MediaTek secure
+	  firmware backend.
+
+config MTK_FW_ENCRYPT_VIA_BL31
+	bool "Decrypt encrypted FIT firmware via BL31"
+	depends on MTK_FW_ENCRYPT
+	help
+	  Use the MediaTek BL31 firmware-decryption SMC service.
 endif
--- a/boot/Makefile
+++ b/boot/Makefile
@@ -55,2 +55,4 @@ obj-$(CONFIG_$(PHASE_)FIT_CIPHER) += image-cipher.o
 obj-$(CONFIG_$(PHASE_)FIT_CIPHER) += image-cipher.o
+obj-$(CONFIG_MTK_FW_ENCRYPT) += mtk_fw_dec.o
+obj-$(CONFIG_MTK_FW_ENCRYPT_VIA_BL31) += mtk_aes_decrypt.o
 obj-$(CONFIG_$(PHASE_)FIT_SIGNATURE) += fdt_region.o
--- a/boot/image-cipher.c
+++ b/boot/image-cipher.c
@@ -49,3 +49,18 @@ struct cipher_algo cipher_algos[] = {
 		.decrypt = image_aes_decrypt,
 		.add_cipher_data = image_aes_add_cipher_data
-	}
+	},
+	{
+		.name = "tee_aes256",
+		.key_len = AES256_KEY_LENGTH,
+		.iv_len = AES_BLOCK_LENGTH,
+#if IMAGE_ENABLE_ENCRYPT
+		.calculate_type = EVP_aes_256_cbc,
+#endif
+		.encrypt = image_aes_encrypt,
+#ifdef USE_HOSTCC
+		.decrypt = image_aes_decrypt,
+#else
+		.decrypt = mtk_image_aes_decrypt,
+#endif
+		.add_cipher_data = image_aes_add_cipher_data
+	}
 };
--- a/include/u-boot/aes.h
+++ b/include/u-boot/aes.h
@@ -44,3 +44,16 @@ int image_aes_decrypt(struct image_cipher_info *info,
 #endif /* IMAGE_ENABLE_DECRYPT */
+
+#if CONFIG_IS_ENABLED(MTK_FW_ENCRYPT_VIA_BL31)
+int mtk_image_aes_decrypt(struct image_cipher_info *info,
+			  const void *cipher, size_t cipher_len,
+			  void **data, size_t *size);
+#else
+static inline int mtk_image_aes_decrypt(struct image_cipher_info *info,
+					const void *cipher, size_t cipher_len,
+					void **data, size_t *size)
+{
+	return -ENXIO;
+}
+#endif
 
 #endif
--- a/arch/arm/lib/bootm.c
+++ b/arch/arm/lib/bootm.c
@@ -23,2 +23,3 @@
 #include <image.h>
+#include <mtk_fw_dec.h>
 #include <u-boot/zlib.h>
@@ -235,2 +236,9 @@ static void boot_prep_linux(struct bootm_headers *images)
 	board_prep_linux(images);
+
+	if (CONFIG_IS_ENABLED(MTK_FW_ENCRYPT)) {
+		int ret = fw_dec_cleanup();
+
+		if (ret)
+			panic("Cannot clean up firmware encryption keys\n");
+	}
 }
